test/test-mongocrypt-crypto.c

```diff
@@ -19,325 +19,304 @@
+      goto done;
+   }
+   ASSERT_OK_STATUS (ret, status);
+   out.len = outlen;
+   ASSERT_CMPBUF (out, test->plaintext);
+   out.len = test->algo->get_plaintext_len (test->ciphertext.len, status);
+
+   // Negative: Mutated IV
+   _mongocrypt_buffer_t modified_ciphertext = {0};
+   _mongocrypt_buffer_copy_to (&test->ciphertext, &modified_ciphertext);
+   _mongocrypt_buffer_resize (&out, plaintext_len);
+   modified_ciphertext.data[0] ^= 1;
+   ret = test->algo->do_decrypt (crypt->crypto,
+                                 &test->aad,
+                                 &test->key,
+                                 &modified_ciphertext,
+                                 &out,
+                                 &outlen,
+                                 status);
+   out.len = outlen;
+   ASSERT_BAD_DECRYPT (ret, out, test, status);
+
+   // Negative: Mutated ciphertext
+   _mongocrypt_buffer_copy_to (&test->ciphertext, &modified_ciphertext);
+   _mongocrypt_buffer_resize (&out, plaintext_len);
+   modified_ciphertext.data[MONGOCRYPT_IV_LEN] ^= 1;
+   ret = test->algo->do_decrypt (crypt->crypto,
+                                 &test->aad,
+                                 &test->key,
+                                 &modified_ciphertext,
+                                 &out,
+                                 &outlen,
+                                 status);
+   ASSERT_BAD_DECRYPT (ret, out, test, status);
+
+   // Negative: Mutated tag
+   // Note: On algorithms without HMAC, this just repeats the mutated ciphertext
+   // test in a differnt part of S.
+   _mongocrypt_buffer_copy_to (&test->ciphertext, &modified_ciphertext);
+   _mongocrypt_buffer_resize (&out, plaintext_len);
+   modified_ciphertext.data[modified_ciphertext.len - 1] ^= 1;
+   ret = test->algo->do_decrypt (crypt->crypto,
+                                 &test->aad,
+                                 &test->key,
+                                 &modified_ciphertext,
+                                 &out,
+                                 &outlen,
+                                 status);
+   ASSERT_BAD_DECRYPT (ret, out, test, status);
+
+   _mongocrypt_buffer_cleanup (&modified_ciphertext);
+done:
+   _mongocrypt_buffer_cleanup (&out);
    mongocrypt_status_destroy (status);
    mongocrypt_destroy (crypt);
+
+   printf ("End %s...\n", test->name);
 }
 
-typedef struct {
-   const char *testname;
-   const char *key;
-   const char *iv;
-   const char *plaintext;
-   const char *ciphertext;
-} aes_256_ctr_test_t;
+static const _mongocrypt_value_encryption_algorithm_t *
+get_algo_by_name (const char *name)
+{
+   if (!strcmp (name, "AES-256-CBC/SHA-512-256") || !strcmp (name, "FLE1")) {
+      return _mcFLE1Algorithm ();
+   }
+   if (!strcmp (name, "AES-256-CTR/SHA-256") || !strcmp (name, "FLE2")) {
+      return _mcFLE2AEADAlgorithm ();
+   }
+   if (!strcmp (name, "AES-256-CTR/NONE") || !strcmp (name, "FLE2IEV")) {
```

Comment:
```suggestion
   if (!strcmp (name, "AES-256-CTR/NONE") || !strcmp (name, "FLE2AEAD")) {
```

Following suggestion in https://github.com/mongodb/libmongocrypt/pull/580#discussion_r1116497145